### PR TITLE
NIFI-7371 - Improve error handling of S3 processors

### PR DESCRIPTION
**Pull request by greyp9**
Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR
Adds exception attributes to flow file on failure of S3 processor operations.  New attributes are derived from information contained in the S3 exception class.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [X] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [X] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [X] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [X] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [X] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [X] Have you written or updated unit tests to verify your changes?
- [X] Have you verified that the full build is successful on JDK 8?
- [X] Have you verified that the full build is successful on JDK 11?
- [-] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [-] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [-] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [-] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by greyp9** [comment]
Pulled these attributes out of a NiFi instance testing some failure conditions:

[INCORRECT BUCKET NAME]
s3.additionalDetails = {BucketName=<bad bucket name>, Error=<base64 string>}
s3.errorCode = NoSuchBucket
s3.errorMessage = The specified bucket does not exist
s3.exception = com.amazonaws.services.s3.model.AmazonS3Exception
s3.statusCode = 404

[BAD FILENAME] (appended garbage to end of FetchS3Object field "Object Key")
s3.additionalDetails = {Error=<base64 string>, Key=<bad path>}
s3.errorCode = NoSuchKey
s3.errorMessage = The specified key does not exist.
s3.exception = com.amazonaws.services.s3.model.AmazonS3Exception
s3.statusCode = 404

[NETWORK CONNECTIVITY FAILURE] (disabled wifi)
s3.exception = com.amazonaws.SdkClientException

[BAD CREDENTIALS] (incorrect AWS Secret Access Key)
s3.additionalDetails = {CanonicalRequestBytes=<hex string>}
s3.errorCode = SignatureDoesNotMatch
s3.errorMessage = The request signature we calculated does not match the signature you provided. Check your key and signing method.
s3.exception = com.amazonaws.services.s3.model.AmazonS3Exception
s3.statusCode = 403

[BAD CREDENTIALS] (incorrect AWS Access Key ID)
s3.additionalDetails = {AWSAccessKeyId=<bad key id>, Error=<base64 string>}
s3.errorCode = InvalidAccessKeyId
s3.errorMessage = The AWS Access Key Id you provided does not exist in our records.
s3.exception = com.amazonaws.services.s3.model.AmazonS3Exception
s3.statusCode = 403

File: nifi-nar-bundles/nifi-aws-bundle/nifi-aws-abstract-processors/src/main/java/org/apache/nifi/processors/aws/s3/AbstractS3Processor.java [hunk 1]
```diff
@@ -315,6 +318,23 @@ protected final AccessControlList createACL(final ProcessContext context, final
...
 
+    protected FlowFile extractExceptionDetails(final Exception e, final ProcessSession session, FlowFile flowFile) {
+        flowFile = session.putAttribute(flowFile, "s3.exception", e.getClass().getName());
+        if (e instanceof AmazonS3Exception) {
+            flowFile = putAttribute(session, flowFile, "s3.additionalDetails", ((AmazonS3Exception) e).getAdditionalDetails());
+        }
+        if (e instanceof AmazonServiceException) {
+            flowFile = putAttribute(session, flowFile, "s3.statusCode", ((AmazonServiceException) e).getStatusCode());
+            flowFile = putAttribute(session, flowFile, "s3.errorCode", ((AmazonServiceException) e).getErrorCode());
+            flowFile = putAttribute(session, flowFile, "s3.errorMessage", ((AmazonServiceException) e).getErrorMessage());
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-aws-bundle/nifi-aws-abstract-processors/src/main/java/org/apache/nifi/processors/aws/s3/AbstractS3Processor.java:329]: Just a minor adjustment, it looks like the exception could be cast once as an `AmazonServiceException` and then the various methods could be called.
```suggestion
            final AmazonServiceException ase = (AmazonServiceException) e;
            flowFile = putAttribute(session, flowFile, "s3.statusCode", ase.getStatusCode());
            flowFile = putAttribute(session, flowFile, "s3.errorCode", ase.getErrorCode());
            flowFile = putAttribute(session, flowFile, "s3.errorMessage", ase.getErrorMessage());
```

**Review by exceptionfactory — COMMENTED** [review]
Thanks for the improvement @greyp9, the additional attributes should support a variety of use cases in handling failure scenarios. I noted one minor optimization, otherwise it looks good.

File: nifi-nar-bundles/nifi-aws-bundle/nifi-aws-abstract-processors/src/main/java/org/apache/nifi/processors/aws/s3/AbstractS3Processor.java (hunk 1 above)
Comment (greyp9) [reply nifi-nar-bundles/nifi-aws-bundle/nifi-aws-abstract-processors/src/main/java/org/apache/nifi/processors/aws/s3/AbstractS3Processor.java]: Sure; update pushed.

**Review by exceptionfactory — APPROVED** [review]
Thanks for the updates @greyp9, looks good! +1 Merging.

**Pull request closed without merging**